File: integrasjonspunkt/src/main/java/no/difi/meldingsutveksling/nextmove/ResponseStatusSender.java
package no.difi.meldingsutveksling.nextmove;

import no.difi.meldingsutveksling.ServiceIdentifier;
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties;
import no.difi.meldingsutveksling.domain.sbdh.StandardBusinessDocument;
import no.difi.meldingsutveksling.receipt.ReceiptStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class ResponseStatusSender {

    private final ResponseStatusSenderProxy proxy;
    private final IntegrasjonspunktProperties props;
    private final Logger log = LoggerFactory.getLogger(ResponseStatusSender.class);

    public ResponseStatusSender(ResponseStatusSenderProxy proxy, IntegrasjonspunktProperties props) {
        this.proxy = proxy;
        this.props = props;
    }

    public void queue(StandardBusinessDocument sbd, ServiceIdentifier si, ReceiptStatus status) {
        if (props.getNextmove().getStatusServices().contains(si)) {
            Thread t = new Thread(() -> {
                try {
                    proxy.queue(sbd, si, status);
                } catch (Throwable t1) {
                    log.error("Error sending status message", t1);
                }
            }, "response-status-sender");
            t.setDaemon(true);
            t.start();
        }
    }

}

File: integrasjonspunkt/src/main/kotlin/no/difi/meldingsutveksling/nextmove/ResponseStatusSender.kt
package no.difi.meldingsutveksling.nextmove

import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import no.difi.meldingsutveksling.ServiceIdentifier
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties
import no.difi.meldingsutveksling.domain.sbdh.StandardBusinessDocument
import no.difi.meldingsutveksling.receipt.ReceiptStatus
import no.difi.meldingsutveksling.sbd.SBDFactory
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Lazy
import org.springframework.retry.annotation.Backoff
import org.springframework.retry.annotation.Retryable
import org.springframework.stereotype.Component

@Component
class ResponseStatusSender(
    val proxy: ResponseStatusSenderProxy,
    val props: IntegrasjonspunktProperties
) {
    val log = LoggerFactory.getLogger(ResponseStatusSender::class.java)

    fun queue(sbd: StandardBusinessDocument, si: ServiceIdentifier, status: ReceiptStatus) {
        if (si in props.nextmove.statusServices) {
            CoroutineScope(Dispatchers.IO).launch(CoroutineExceptionHandler { _, t -> log.error("Error sending status message", t) }) {
                proxy.queue(sbd, si, status)
            }
        }
    }

}

@Component
open class ResponseStatusSenderProxy(
    @Lazy val internalQueue: InternalQueue,
    private val sbdFactory: SBDFactory
) {

    @Retryable(maxAttempts = 10, backoff = Backoff(delay = 5000, multiplier = 2.0, maxDelay = 1000 * 60 * 10L))
    open fun queue(sbd: StandardBusinessDocument, si: ServiceIdentifier, status: ReceiptStatus) {
        sbdFactory.createStatusFrom(sbd, status)
            ?.let { NextMoveOutMessage.of(it, si) }
            ?.let(internalQueue::enqueueNextMove)
    }
}
